[clang] Add __nullptr as a keyword to C

AidanGoldfarb opened the pull request:
This PR resolves #121503. 

I was not sure whether I added my test in the correct place, and if so, if that deprecates [nullptr-98.cpp](https://github.com/llvm/llvm-project/blob/main/clang/test/SemaCXX/nullptr-98.cpp)

Commit:
__nullptr -> KEYALL, added relevant test

Sirraide commented on clang/test/Sema/nullptr.c:
```diff
@@ -108,3 +108,6 @@ void test_f1() {
 
+// __nullptr keyword in C
+void foo(void *);
+void bar() { foo(__nullptr); }
```
nit: missing newline at the end of the file

Sirraide commented on clang/test/Sema/nullptr.c:
```diff
@@ -108,3 +108,6 @@ void test_f1() {
   int ir = (f1)(nullptr);
 }
 
+// __nullptr keyword in C
```
Maybe also add these:
```c
static_assert(nullptr == __nullptr);
static_assert(_Generic(typeof(__nullptr), nullptr_t: true, default: false));
```
Just so we have a few more checks. I don’t think we need to go as far as to copy-paste the rest of this file for this.

(I don’t write these `_Generic` selections too often so I hope I got that one right.)

Sirraide reviewed (commented):
Looking good overall, but this still needs a release note.

AidanGoldfarb committed:
Added additional tests

static_assert(nullptr == __nullptr) and static_assert(_Generic(typeof(__nullptr), nullptr_t: true, default: false))

AaronBallman reviewed (commented):
Please be sure to also update `clang/docs/ReleaseNotes.rst` so users know about the new functionality. Also, not signing you up for this work, but we really should document all of these keywords in `clang/docs/LanguageExtensions.rst`. Maybe it's worth starting with this one and hopefully we can backfill the rest in the future?

AidanGoldfarb committed:
Update nullptr.c

AidanGoldfarb committed:
Update ReleaseNotes.rst

AidanGoldfarb committed:
Update nullptr.c

Removed duplicate test

erichkeane commented on clang/docs/ReleaseNotes.rst:
```diff
@@ -325,6 +325,7 @@ C Language Changes
 ------------------
 
 - Extend clang's ``<limits.h>`` to define ``LONG_LONG_*`` macros for Android's bionic.
+- Added support for the ``__nullptr`` keyword. 
```
```suggestion
- Exposed the the ``__nullptr`` keyword as an alias for ``nullptr`` in all C language modes.
```

Perhaps??

AidanGoldfarb committed:
Update clang/docs/ReleaseNotes.rst

Co-authored-by: Erich Keane <ekeane@nvidia.com>

erichkeane commented on clang/test/Sema/nullptr.c:
```diff
@@ -108,3 +108,8 @@ void test_f1() {
+void foo(void *);
+void bar() { foo(__nullptr); }
+static_assert(nullptr == __nullptr);
+static_assert(_Generic(typeof(__nullptr), nullptr_t: true, default: false));
```
This isn't actually done here.  This test is C2x only, Aaron was asking for a C17 test as well.

Commit:
added c17 tests

erichkeane commented on clang/test/Sema/nullptr.c:
```diff
@@ -113,5 +113,3 @@ void foo(void *);
 void bar() { foo(__nullptr); }
 static_assert(nullptr == __nullptr);
 static_assert(_Generic(typeof(__nullptr), nullptr_t: true, default: false));
-_Static_assert(__nullptr == 0); // Test that its value matches that of NULL
```
I would probably leave the tests from aaron in BOTH modes, but otherwise LGTM.

Commit:
duplicated c17 tests in c2x

AidanGoldfarb commented:
Updating LanguageExtensions.rst before merge (and final reviews).

Commit:
updated LanguageExtension.rst

Commit:
updated LanguageExtension.rst

Commit:
updated LanguageExtension.rst

AidanGoldfarb committed:
Fixed ReleaseNotes.rst

AaronBallman commented on clang/docs/LanguageExtensions.rst:
```diff
@@ -514,9 +514,7 @@ available in all language modes.
-``__nullptr`` is an alternate spelling for ``nullptr``, but is also available in
-C++ modes prior to C++11. Note that it's currently not availbale in C despite
-C23 having support for ``nullptr``.
+``__nullptr`` is an alternate spelling for ``nullptr``, it is available since C++98 and C23.
```
```suggestion
``__nullptr`` is an alternate spelling for ``nullptr``. It is available in all C and C++ language modes.
```

AaronBallman reviewed (approved):
LGTM aside from a change to the documentation. Thank you!

AidanGoldfarb committed:
Update clang/docs/LanguageExtensions.rst

Co-authored-by: Aaron Ballman <aaron@aaronballman.com>
